spec/upload_to_s3_spec.rb

```diff
@@ -145,39 +145,141 @@ def stub_s3_response_for_file(key, exists: true)
+              file: file_path,
+              if_exists: :fail
+            )
+          end.to raise_error(FastlaneCore::Interface::FastlaneError, "File already exists in S3 bucket #{test_bucket} at #{expected_key}")
+        end
+      end
+
+      it 'logs a message without failing when if_exists is :skip' do
+        expected_key = '29d5f92e9ee44d4854d6dfaeefc3dc27d779fdf3/existing-key'
+        stub_s3_response_for_file(expected_key)
+
+        warnings = []
+        allow(FastlaneCore::UI).to receive(:important) { |message| warnings << message }
+
+        with_tmp_file(named: 'existing-key') do |file_path|
+          key = run_described_fastlane_action(
+            bucket: test_bucket,
+            key: 'existing-key',
+            file: file_path,
+            if_exists: :skip
+          )
+
+          expect(warnings).to eq(["File already exists in S3 bucket #{test_bucket} at #{expected_key}. Skipping upload."])
+          expect(key).to eq(expected_key)
+        end
+      end
+
+      it 'upload the file overriding the existing one when if_exists is :replace' do
+        expected_key = '29d5f92e9ee44d4854d6dfaeefc3dc27d779fdf3/existing-key'
+        stub_s3_response_for_file(expected_key)
+
+        warnings = []
+        allow(FastlaneCore::UI).to receive(:important) { |message| warnings << message }
+
+        with_tmp_file(named: 'existing-key') do |file_path|
+          expect(client).to receive(:put_object).with(body: file_instance_of(file_path), bucket: test_bucket, key: expected_key)
+
+          return_value = run_described_fastlane_action(
+            bucket: test_bucket,
+            key: 'existing-key',
+            file: file_path,
+            if_exists: :replace
+          )
+
+          expect(return_value).to eq(expected_key)
+          expect(Fastlane::Actions.lane_context[Fastlane::Actions::SharedValues::S3_UPLOADED_FILE_PATH]).to eq(expected_key)
+        end
+      end
+
+      it 'throws when if_exists is not one of the expected values' do
+        with_tmp_file(named: 'key') do |file_path|
+          expect do
+            run_described_fastlane_action(
+              bucket: test_bucket,
+              key: 'a8c-key1',
+              file: file_path,
+              if_exists: :invalid
+            )
+          end.to raise_error(FastlaneCore::Interface::FastlaneError, '`if_exist` must be one of :skip, :replace, :fail')
+        end
+      end
+
+      it 'does not allow to have both skip_if_exists and if_exists set at the same time' do
+        with_tmp_file(named: 'existing-key') do |file_path|
+          expect do
+            run_described_fastlane_action(
+              bucket: test_bucket,
+              key: 'existing-key',
+              file: file_path,
+              if_exists: :skip,
+              skip_if_exists: true
+            )
+          end.to raise_error(FastlaneCore::Interface::FastlaneError, 'You cannot set both :skip_if_exists and :if_exists. Please only use :if_exists.')
```

Comment:
💄  Nitpick: wdyt about trying **_all_** non-nil combinations, instead of just a specific arbitrary one? 🤔 

```suggestion
          [true, false].product(%i[skip fail replace]).each do |skip_if_exists, if_exists|
            expect do
              run_described_fastlane_action(
                bucket: test_bucket,
                key: 'existing-key',
                file: file_path,
                if_exists: if_exists,
                skip_if_exists: skip_if_exists
              )
            end.to raise_error(FastlaneCore::Interface::FastlaneError, 'You cannot set both :skip_if_exists and :if_exists. Please only use :if_exists.')
          end
```

Reply:
Brilliant!

This, too, is a bit on the edge of ours vs Fastlane's behavior. However, I think the critical difference here is that for this behavior to take place we need to configure Fastlane explicitly, whereas in the non-`Symbol` conversion it was all automatic

Reply:
> This, too, is a bit on the edge of ours vs Fastlane's behavior. However, I think the critical difference here is that for this behavior to take place we need to configure Fastlane explicitly,

Totally. In particular, I think that this test doesn't just verify that we have set up `conflicting_options` correctly, but also checks that we have decided — via the `conflict_block` customization — to reject _any_ combination of those. As opposed to e.g. allowing the not-so-conflicting options like `skip_if_exists: true, if_exists: :skip` or `skip_if_exists: false, if_exists: :fail` for example.

So this test is really to verify that we won't have any future regressions about "rejecting any non-nil combination", even if decide to change our implementation of `conflict_block` in the future 🙂

Reply:
Addressed in 97999706685ce0ba933aec98b589cd759ee21438